### PR TITLE
Update error span for attribute / data kind mismatches

### PR DESCRIPTION
**Pull request by jplatte**
Follow-up to https://github.com/serde-rs/serde/pull/1695#discussion_r1275423042.

There are a few more in line 770 and below, but there it wasn't quite as clear what to set the span to.

### CODE REVIEW

File: test_suite/tests/ui/default-attribute/enum_path.stderr [hunk 1]
```diff
@@ -1,5 +1,5 @@
 error: #[serde(default = "...")] can only be used on structs with named fields
- --> tests/ui/default-attribute/enum_path.rs:5:1
+ --> tests/ui/default-attribute/enum_path.rs:4:9
   |
-5 | enum E {
-  | ^^^^
+4 | #[serde(default = "default_e")]
+  |         ^^^^^^^
```
Comment (dtolnay) [inline test_suite/tests/ui/default-attribute/enum_path.stderr:5]: I am not sure whether to prefer underlining just the ident or the whole nested attribute. Do you have an opinion?

```console
#[serde(before, default = "...", after)]
                ^^^^^^^
```

```console
#[serde(before, default = "...", after)]
                ^^^^^^^^^^^^^^^
```

I think the second behavior is implementable with `cx.syn_error(meta.error(msg))`.

**Review by dtolnay — APPROVED** [review]
Thanks!

File: test_suite/tests/ui/default-attribute/enum_path.stderr (hunk 1 above)
Comment (jplatte) [reply test_suite/tests/ui/default-attribute/enum_path.stderr:5]: I guess the latter is a bit nicer, thanks for the impl hint.

**Review by dtolnay — APPROVED** [review]
Looks good! Thank you.

**Pull request merged**